### PR TITLE
Use use-context-selector versions of createContext, useContext

### PR DESCRIPTION
**Pull request by cvolzke4**
## Summary

Switch to using createContext, useContext (without adding selectors just yet) from use-context-selector as per go/story-editor-context-perf:
https://docs.google.com/document/d/18m9HamHkOvvTrv-mSOJawha0K4KTdn49HrnfXShULDI/edit#heading=h.21vtkbcpsflf

This will allow specific attributes of the context state to be selected without re-rendering after every context state change.

React profiler of scrolling through media after this series of changes:
https://share.getcloudapp.com/YEu14qDg

Next PR:
https://github.com/google/web-stories-wp/pull/1705

## To-do

Next steps:
- Add selector functions to useContextSelectorShallow calls.
- Fix recursion between useMemo and useUploader (pre-existing bug).
- Add useMemo at the end of each context provider.
- Use Pure Components for simple components.
- Prevent action functions from depending on state changes.

---

Fixes #1466

### CODE REVIEW

File: assets/src/edit-story/app/api/useAPI.js [hunk 1]
```diff
@@ -14,18 +14,14 @@
...
 /**
  * Internal dependencies
  */
+import { identity, useContextSelectorShallow } from '../../utils/context';
 import Context from './context';
 
-function useAPI() {
-  return useContext(Context);
+function useAPI(selector) {
+  return useContextSelectorShallow(Context, selector ?? identity);
```
Comment (dvoytenko) [inline assets/src/edit-story/app/api/useAPI.js]: Hmm. This is an interesting twist. Would we allow a selector function to change ever? Does the framework make it changeable or is only the first one used? Otherwise, I feel like the `selector ?? identity` will be used a lot, so I think we need to push it back to the implementation of the `useContextSelectorShallow`. As an optimization, when `identity` is used, it also means you actually don't need to do any shallow comparison. I don't know what the library we have does. Perhaps they already do `if (a === b) return true`, but I think being explicit would be helpful.

File: assets/src/edit-story/app/history/context.js [hunk 2]
```diff
@@ -15,8 +15,8 @@
  */
 
 /**
- * External dependencies
+ * Internal dependencies
```
Comment (dvoytenko) [inline assets/src/edit-story/app/history/context.js:18]: Up to you, but do we want to go with every context use? Or should we first merge 2-3 key ones and then expand?

File: assets/src/edit-story/app/uploader/useUploader.js [hunk 3]
```diff
@@ -29,13 +29,20 @@ import { __, sprintf } from '@wordpress/i18n';
...
 import { useAPI } from '../../app/api';
 import { useConfig } from '../config';
-import { useMedia } from '../media';
 import createError from '../../utils/createError';
 
 function useUploader(refreshLibrary = true) {
-  const {
-    actions: { resetWithFetch },
-  } = useMedia();
+  // TODO(https://github.com/google/web-stories-wp/issues/1703):
```
Comment (dvoytenko) [inline assets/src/edit-story/app/uploader/useUploader.js]: This feels like it needs to go into its own pull request.

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: assets/src/edit-story/app/uploader/useUploader.js (hunk 3 above)
Comment (cvolzke4) [reply assets/src/edit-story/app/uploader/useUploader.js]: Done: https://github.com/google/web-stories-wp/pull/1799

File: assets/src/edit-story/app/api/useAPI.js (hunk 1 above)
Comment (cvolzke4) [reply assets/src/edit-story/app/api/useAPI.js]: The selector function can change for any call to useContextSelector, so two calls to useContextSelector can select different fragments.

In the final changes, 'selector' will be a mandatory argument for useMedia, useStory and useCanvas where performance is more critical, but optional for useAPI, useConfig etc where changes happen infrequently (or not at all).

File: assets/src/edit-story/app/history/context.js (hunk 2 above)
Comment (cvolzke4) [reply assets/src/edit-story/app/history/context.js:18]: I was thinking that we could do them all in one go, since it's a low-risk change and all the same type of change (so it's easy to code review).

If we do find problems, useContext/useContextSelectorShallow can be trivially modified call the React version to verify if this is the cause.

File: assets/src/edit-story/app/api/useAPI.js (hunk 1 above)
Comment (dvoytenko) [reply assets/src/edit-story/app/api/useAPI.js]: Sounds good.

**Review by dvoytenko — APPROVED** [review]
